### PR TITLE
Vtwapsubscriber

### PR DESCRIPTION
**Pull request by JamesBurrows**
Vtwapsubscriber is a realtime subscriber using TorQ functionality to connect to tickerplant, subscribe to tables and syms, and build its quote/trade tables. It also has functions to calculate TWAP and VWAP that the client will call with the relevant syms, start-time, and end-time frames.

It has been added to the process csv file under metrics, as it is a metrics related process. This will allow it to connect to TorQ processes and subscribe.

### CODE REVIEW

File: code/processes/vtwap_subscriber.q [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+schema:@[value;`schema;1b];                                                                     //retrieve the schema from the tp
+subscribeto:@[value;`subscribeto;`];
+ignorelist:@[value;`ignorelist;`heartbeat`logmsg];
+subscribesyms:@[value;`subscribesyms;`];
+tpconnsleepintv:@[value;`tpconnsleepintv;10];                                                   //number of seconds between attempts to connect to the rdb
+if[not .timer.enabled;.lg.e[`symcheckinit;
+   "the timer must be enabled to run the vwapsub process"]];                                    //if the timer is not enabled, then exit with error
+subscribe:{[]
+  if[count s:.sub.getsubscriptionhandles[tickerplanttypes;();()!()];
+   subproc:first s;
```
Comment (ThomasSmyth) [inline code/processes/vtwap_subscriber.q]: You can get away with defining `subproc`, it is only used once below so you could jus use `first s` on line 15.

File: code/processes/vtwap_subscriber.q [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+tickerplanttypes:@[value;`tickerplanttypes;`tickerplant];                                       //list of tp types to look for and call in
+schema:@[value;`schema;1b];                                                                     //retrieve the schema from the tp
+subscribeto:@[value;`subscribeto;`];
+ignorelist:@[value;`ignorelist;`heartbeat`logmsg];
+subscribesyms:@[value;`subscribesyms;`];
+tpconnsleepintv:@[value;`tpconnsleepintv;10];                                                   //number of seconds between attempts to connect to the rdb
+if[not .timer.enabled;.lg.e[`symcheckinit;
+   "the timer must be enabled to run the vwapsub process"]];                                    //if the timer is not enabled, then exit with error
+subscribe:{[]
+  if[count s:.sub.getsubscriptionhandles[tickerplanttypes;();()!()];
```
Comment (ThomasSmyth) [inline code/processes/vtwap_subscriber.q]: I think this condition should be that if `0=count s...` then exit with `:()`. Then the rest of the code block gets executed if that is false.

File: code/processes/vtwap_subscriber.q [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+   subproc:first s;
+   .lg.o[`subscribe;"found available tickerplant, attempting to subscribe"];
+   subinfo:.sub.subscribe[subscribeto;subscribesyms;schema;0b;subproc];
+   @[`.vtwap;;:;]'[key subinfo;value subinfo];
```
Comment (ThomasSmyth) [inline code/processes/vtwap_subscriber.q]: Should work without each-both ``@[`.vtwap;key subinfo;:;value subinfo]``

File: code/processes/vtwap_subscriber.q [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  .servers.startup[];                                                                           //run the servers startup code again (to make connection to discovery)
+ ];
+
+calctwap:{[x;st;et] select TWAP:sum[price]%sum[time] by sym from twap where sym in x,minute within (st,et)}
+
+calcvwap:{[x;st;et] select VWAP:sum[price]%sum[size] by sym from vwap where sym in x,minute within (st,et)}
+
+upd:{[t;x]
+  t insert x;
+  if[t in `trade;
```
Comment (ThomasSmyth) [inline code/processes/vtwap_subscriber.q]: Again another condition that can be flipped to exit here instead. ``if[not t in `trade;:()];`` Avoids the need to indent the following block + should this work with quotes and trades?

File: code/processes/vtwap_subscriber.q (hunk 2 above)
Comment (JamesBurrows) [reply code/processes/vtwap_subscriber.q]: It's doing if `count s` is true, then append to out log with "Found tickerplant", isn't it? So if the condition is `0b=count s` then if there is no tickerplant (i.e. `0b=count s`) it'll say it's found?

Comment (ThomasSmyth) [reply code/processes/vtwap_subscriber.q]: My idea is that this should read:
```
if[0=count s:.sub.getsubscriptionhandles[tickerplanttypes;();()!()];:()];
```
So you exit when then count is 0. Its achieves the same thing without having to indent the code block below it.

Comment (JamesBurrows) [reply code/processes/vtwap_subscriber.q]: Understood, thanks

**Comment by ThomasSmyth** [comment]
Please merge.

**Pull request merged**